### PR TITLE
Update endpoint

### PR DESCRIPTION
**Pull request by Aurora1024**
Test Cases for YourResourceModel Model 
- Test create a recommendation
- Test deserialization of a Recommendation
- Test deserialization of bad data
- Test deserialization of a Recommendation with missing data
- Find a recommendation type by two product ids
- Test serialization of a Recommendation
- Update a recommendation type by two product ids
- Update a recommendation type by two product ids

REST API Recommendation Server Tests 
- Create a new recommendation
- Create a Recommendation with missing data
- Create a Recommendation with no content type
- Get a single Recommendation
- Get a Recommendation thats not found
- Test index call
- Update an existing recommendation
- create a Recommendation with no content type
- update a Recommendation that is not found

Name                  Stmts   Miss  Cover   Missing
---------------------------------------------------
service/__init__.py      25      4    84%   30, 39-42
service/models.py        55      0   100%
service/routes.py        52      0   100%
service/status.py        46      0   100%
---------------------------------------------------
TOTAL                   178      4    98%
-----------------------------------------------------------------------------
17 tests run in 0.905 seconds (17 tests passed)

### CODE REVIEW

File: service/routes.py [hunk 1]
```diff
@@ -80,6 +80,33 @@ def get_recommendations(product_id1, product_id2):
...
+    update a relationship
+    This endpoint will update a relationship based the data in the body that is posted
+    """
+    app.logger.info("Request to update a ")
+    check_content_type("application/json")
+    recommendation = Recommendation()
+    recommendation.deserialize(request.get_json())
+    old_recommendation = recommendation.find(product_id1, product_id2)
+    if not old_recommendation:
+        raise BadRequest("No existing relationship between product {} and {}".format(product_id1, product_id2))
```
Comment (zyy9871) [inline service/routes.py]: Should raise NotFound. No idea how to cover this line.

File: service/routes.py [hunk 2]
```diff
@@ -80,6 +80,33 @@ def get_recommendations(product_id1, product_id2):
...
+    if recommendation.relationship == "GO_TOGETHER":
+        old_recommendation.relationship = Type.GO_TOGETHER
+    elif recommendation.relationship == "CROSS_SELL":
+        old_recommendation.relationship = Type.CROSS_SELL
+    elif recommendation.relationship == "UP_SELL":
+        old_recommendation.relationship = Type.UP_SELL
+    elif recommendation.relationship == "ACCESSORY":
+        old_recommendation.relationship = Type.ACCESSORY
```
Comment (zyy9871) [inline service/routes.py]: Just do old_recommendation.relationship = recommendation.relationship

File: service/routes.py (hunk 1 above)
Comment (mandy-cmd) [reply service/routes.py]: To cover this line you could write a unit test trying to update relationship between -1 and -1 and a NOT FOUND as mentioned above should be expected in your test.

File: service/models.py [hunk 3]
```diff
@@ -51,6 +51,15 @@ def create(self):
         db.session.add(self)
         db.session.commit()
 
+    def update(self):
+        """
+        update a recommendation type to the database
+        """
+        if not self.relationship:
```
Comment (mandy-cmd) [inline service/models.py:59]: To test this line, have a test case say test_update_empty_relationship() and do not set relationship in your input. Then DataValidationError should be expected.

File: tests/test_routes.py [hunk 4]
```diff
@@ -128,3 +128,28 @@ def test_create_recommendation_no_content_type(self):
...
+        new_recommendation = resp.get_json()
+        new_recommendation["relationship"] = "CROSS_SELL"
+        logging.debug(new_recommendation)
+        resp = self.app.put(
+            "/recommendations/products/{}/related-products/{}".format(test_recommendation.product_id1,
+                                                                      test_recommendation.product_id2),
+            json=new_recommendation,
+            content_type="application/json"
+        )
+       #new_recommendation["relationship"] = "CROSS_SELL"
```
Comment (mandy-cmd) [inline tests/test_routes.py]: Please remove commented/unused code.

File: service/routes.py (hunk 2 above)
Comment (mandy-cmd) [reply service/routes.py]: Yes old_recommendation.relationship = recommendation.relationship should be sufficient. Also here's also where you are missing coverage but I don't think you have to test every type UPDATE. Making it old_recommendation.relationship = recommendation.relationship should also help with testing coverage.

File: tests/test_models.py [hunk 5]
```diff
@@ -106,4 +106,21 @@ def test_find_recommendation_type(self):
...
+        recommendation.relationship = Type.CROSS_SELL
+        recommendation.update()
+        self.assertIsNot(recommendation, None)
+        self.assertEqual(recommendation.relationship.name, 'CROSS_SELL')
+        recommendations = recommendation.all()
+        self.assertEqual(len(recommendations), 1)
+        self.assertEqual(recommendations[0].product_id1, recommendation.product_id1)
+        self.assertEqual(recommendations[0].product_id2, recommendation.product_id2)
+        self.assertEqual(recommendations[0].relationship, recommendation.relationship)
+
```
Comment (mandy-cmd) [inline tests/test_models.py:133]: To cover DataValidationError, you could create a new test and say here relationship = None, and then dataValidationError should be thrown.

**Review by mandy-cmd — CHANGES_REQUESTED** [review]

File: service/models.py (hunk 3 above)
Comment (Aurora1024) [reply service/models.py:58]: Could you pls help look at test_update_recommendation_no_relation_type(self)
I try to assign null to relation type in json file, but it raises error when deserialize it. A little confused how to test this one

 service/routes.py line 92 in update_recommendations
      recommendation.deserialize(request.get_json())
    service/models.py line 83 in deserialize
      raise DataValidationError(
   DataValidationError: Invalid YourResourceModel: body of request contained bad or no data
   -------------------- >> begin captured logging << --------------------
   root: DEBUG: <enum 'Type'>
   root: DEBUG: {'product_id1': 1, 'product_id2': 7, 'relationship': 'CROSS_SELL'}
   root: DEBUG: {"product_id1": 1, "product_id2": 7, "relationship": " "}
   --------------------- >> end captured logging << ---------------------

File: tests/test_models.py (hunk 5 above)
Comment (Aurora1024) [reply tests/test_models.py:133]: Done in test_model thanks

File: service/routes.py (hunk 1 above)
Comment (Aurora1024) [reply service/routes.py]: done in test_route

File: tests/test_routes.py [hunk 6]
```diff
@@ -148,8 +149,35 @@ def test_update_recommendation(self):
...
+        test = json.dumps({'product_id1': test_recommendation.product_id1, "product_id2": test_recommendation.product_id2, "relationship": " "})
+        logging.debug(test_recommendation.serialize())
+        logging.debug(test)
+        resp = self.app.put(
+            "/recommendations/products/{}/related-products/{}".format(test_recommendation.product_id1,
+                                                                      test_recommendation.product_id2),
+            json=test,
+            content_type="application/json"
+        )
+        self.assertEqual(resp.status_code, status.HTTP_415_UNSUPPORTED_MEDIA_TYPE)
```
Comment (mandy-cmd) [inline tests/test_routes.py]: Ok I see your error. So it's actually the [deserialization error](https://github.com/dev-ops-recommendation/recommendations/blob/main/service/models.py#L83) meaning that input relationship cannot be empty at all, in which case you could remove this test and remove ```if not self.relationship``` because it won't be able to hit here at all because validation error would raise first before it could reach this check. 
```
DataValidationError: Invalid YourResourceModel: body of request contained bad or no data
   -------------------- >> begin captured logging << --------------------
   root: DEBUG: {'product_id1': 3, 'product_id2': 8, 'relationship': 'CROSS_SELL'}
   root: DEBUG: {"product_id1": 3, "product_id2": 8, "relationship": " "}
   --------------------- >> end captured logging << ---------------------
```

File: service/models.py [hunk 7]
```diff
@@ -51,6 +51,15 @@ def create(self):
...
+            raise DataValidationError("Update called with empty relationship")
```
Comment (mandy-cmd) [inline service/models.py:60]: What I mean is that I think you could remove these two lines and remove no relation tests.

File: tests/test_routes.py (hunk 6 above)
Comment (Aurora1024) [reply tests/test_routes.py]: Yeah. I test this in test_model part. "if not recommendation" is implement in model.py

File: service/models.py (hunk 7 above)
Comment (Aurora1024) [reply service/models.py:60]: Right, it makes sense. But I think I can leave here for double check.

Comment (mandy-cmd) [reply service/models.py:60]: Then you want to test it in test_model.py because route will never hit here if input relationship is empty.

Comment (Aurora1024) [reply service/models.py:60]: already commit new version that test it in  test_model.py . Thanks

**Review by mandy-cmd — APPROVED** [review]
